### PR TITLE
ENH+RF: tools/ci/install-annex.sh helper and switch to use conda build

### PR DESCRIPTION
**Pull request by yarikoptic**
standalone build is slow, so I think it might make sense to switch to conda-forge build.

Most likely it would hang as is now due to a known issue with recent annex.  I might look into making it possible to install specific version from conda-forge -- want first to see if the whole machinery would work. 

Also with that script it should be possible to quickly deploy desired flavor (I guess I should have used that word instead of "scenario") of git-annex build.  Eventually might also figure out how to automate fetching the snapshot build from the most recent one in datalad-extensions.

TODOs @yarikoptic sees before merging
- [X] verify that actually setting TMPDIR was in effect (e.g. pidlock was engaged in nfs based test case)
   Update: Checked as done based on (perhaps unsound) reasoning at https://github.com/datalad/datalad/pull/4640#issuecomment-658245569
- [x] move tune up of ~/.bashrc inside that helper (as cmdline option).  I think we should avoid such actions to be done in .travis.yml, so we could reuse script across CIs etc.

### CODE REVIEW

**Comment by yarikoptic** [comment]
Wow it went down in flames, while testing ora special remote
```
    fsck downloaded object:                          OK (1.05s)
    retrieveKeyFile resume from 33%:                 OK (3.51s)
    fsck downloaded object:                          Traceback (most recent call last):
  File "/home/travis/virtualenv/python3.5.6/bin/git-annex-remote-ora", line 7, in <module>
    from datalad.distributed.ora_remote import main
  File "/home/travis/virtualenv/python3.5.6/lib/python3.5/site-packages/datalad/__init__.py", line 47, in <module>
    cfg = ConfigManager()
  File "/home/travis/virtualenv/python3.5.6/lib/python3.5/site-packages/datalad/config.py", line 259, in __init__
    self.reload(force=True)
  File "/home/travis/virtualenv/python3.5.6/lib/python3.5/site-packages/datalad/config.py", line 316, in reload
    stdout, stderr = self._run(run_args, log_stderr=True)
  File "/home/travis/virtualenv/python3.5.6/lib/python3.5/site-packages/datalad/config.py", line 606, in _run
    out = self._runner.run(self._config_cmd + args, **kwargs)
  File "/home/travis/virtualenv/python3.5.6/lib/python3.5/site-packages/datalad/cmd.py", line 1114, in run
    *args, **kwargs)
  File "/home/travis/virtualenv/python3.5.6/lib/python3.5/site-packages/datalad/cmd.py", line 883, in run
    stdin=stdin)
  File "/opt/python/3.5.6/lib/python3.5/subprocess.py", line 676, in __init__
    restore_signals, start_new_session)
  File "/opt/python/3.5.6/lib/python3.5/subprocess.py", line 1228, in _execute_child
    restore_signals, start_new_session, preexec_fn)
OSError: [Errno 12] Cannot allocate memory
OK (3.60s)
    retrieveKeyFile resume from 0:                   Traceback (most recent call last):
  File "/home/travis/virtualenv/python3.5.6/bin/git-annex-remote-ora", line 7, in <module>
    from datalad.distributed.ora_remote import main
  File "/home/travis/virtualenv/python3.5.6/lib/python3.5/site-packages/datalad/__init__.py", line 47, in <module>
    cfg = ConfigManager()
  File "/home/travis/virtualenv/python3.5.6/lib/python3.5/site-packages/datalad/config.py", line 259, in __init__
    self.reload(force=True)
  File "/home/travis/virtualenv/python3.5.6/lib/python3.5/site-packages/datalad/config.py", line 316, in reload
    stdout, stderr = self._run(run_args, log_stderr=True)
  File "/home/travis/virtualenv/python3.5.6/lib/python3.5/site-packages/datalad/config.py", line 606, in _run
    out = self._runner.run(self._config_cmd + args, **kwargs)
  File "/home/travis/virtualenv/python3.5.6/lib/python3.5/site-packages/datalad/cmd.py", line 1114, in run
    *args, **kwargs)
  File "/home/travis/virtualenv/python3.5.6/lib/python3.5/site-packages/datalad/cmd.py", line 883, in run
    stdin=stdin)
  File "/opt/python/3.5.6/lib/python3.5/subprocess.py", line 676, in __init__
    restore_signals, start_new_session)
  File "/opt/python/3.5.6/lib/python3.5/subprocess.py", line 1228, in _execute_child
    restore_signals, start_new_session, preexec_fn)
OSError: [Errno 12] Cannot allocate memory
FAIL
      Exception: external special remote protocol error, unexpectedly received "" (unable to parse command)
    fsck downloaded object:                          Traceback (most recent call last):
  File "/home/travis/virtualenv/python3.5.6/bin/git-annex-remote-ora", line 7, in <module>
    from datalad.distributed.ora_remote import main
  File "/home/travis/virtualenv/python3.5.6/lib/python3.5/site-packages/datalad/__init__.py", line 47, in <module>
    cfg = ConfigManager()
  File "/home/travis/virtualenv/python3.5.6/lib/python3.5/site-packages/datalad/config.py", line 259, in __init__
    self.reload(force=True)
  File "/home/travis/virtualenv/python3.5.6/lib/python3.5/site-packages/datalad/config.py", line 316, in reload
    stdout, stderr = self._run(run_args, log_stderr=True)
  File "/home/travis/virtualenv/python3.5.6/lib/python3.5/site-packages/datalad/config.py", line 606, in _run
    out = self._runner.run(self._config_cmd + args, **kwargs)
  File "/home/travis/virtualenv/python3.5.6/lib/python3.5/site-packages/datalad/cmd.py", line 1114, in run
    *args, **kwargs)
  File "/home/travis/virtualenv/python3.5.6/lib/python3.5/site-packages/datalad/cmd.py", line 883, in run
    stdin=stdin)
  File "/opt/python/3.5.6/lib/python3.5/subprocess.py", line 676, in __init__
    restore_signals, start_new_session)
  File "/opt/python/3.5.6/lib/python3.5/subprocess.py", line 1228, in _execute_child
    restore_signals, start_new_session, preexec_fn)
OSError: [Errno 12] Cannot allocate memory
FAIL
      Exception: external special remote protocol error, unexpectedly received "" (unable to parse command)
```
need to sweep in local conda env

**Comment by yarikoptic** [comment]
I wonder if this is the same issue which causes github workflow hang and show no longs when running with recent git annex and master branch: https://github.com/datalad/datalad-extensions/issues/10 . It also feels like might be something python version (3.5 is to EOL later this year) specific.  I will next try with a more recent python

**Comment by yarikoptic** [comment]
eh -- the same with 3.7 -- so it is not it. Also I do not see a button to stop that hanged build on travis any longer :-(  I will check if there is some ready to use nose helper to time out hanging tests

**Comment by yarikoptic** [comment]
woooo -- builtin `@nose.tools.timed`: https://nose.readthedocs.io/en/latest/testing_tools.html . will add it for that test now

**Comment by yarikoptic** [comment]
FWIW -- could not time locally ATM due to https://github.com/datalad/datalad/issues/4643 filed separately... will downgrade annex to known to work ok version...

**Comment by yarikoptic** [comment]
d'oh -- pushed c41a25e788114025ceeb23447198bc466f4d63f6 which introduced use of `@timed` but only then mentioned [comment](https://stackoverflow.com/questions/23625858/timeout-on-tests-with-nosetests#comment51445055_23634666) it just fails the test if it took too long but it doesn't interrupt hanged on.  So - useless AFAIK.  I will let this single run complete though. Next will try `--process-timeout=360 --processes=1` - I think not a single test should run over 5 minutes. I will do a local sweep with a timeout of 120 meanwhile.

**Comment by yarikoptic** [comment]
eh - nope.  Even with `--processes=1 --process-timeout=120 --process-restartworker` 120 is the total tests run time allotted.  So as soon as it hits 2 minutes, starts failing all the rest :-/

**Comment by kyleam** [comment]
Thanks for working on this.  The install script looks like a nice addition.

I'm about to push an update here.  The only code change is to make the PATH adjustment seen by ssh commands.  I've rebased to deal with conflicts in .travis.yml.  I'm switching to maint as the base because (1) this is only related to our testing setup, (2) I think ideally maint and master should share a similar setup if possible, and (3) it'd be nice for maint's tests to be sped up too.

I've dropped the temporary commit that removed other testing because a [test run][0] with the changes here passed.  (And, at least based on that one run, it looks like a good speed up.)

[0]: https://travis-ci.org/github/datalad/datalad/builds/703681374

<details>
<summary>range-diff</summary>

```
1:  c62a768a5 = 1:  fc973c2d5 NF: a script to centralize installation of git-annex in various scenarios
2:  848a9e679 = 2:  a462a4dcd RF: .travis.yml - install annex using script, switch to conda-forge build
3:  5e65dbd38 = 3:  8e0ae333c RM: install-annex-snapshot.sh - just use install-annex.sh snapshot from now on
4:  1fed2f498 < -:  --------- TEMP: remove all matrixes etc
5:  358d37618 = 4:  fd067d1e8 ENH: make it possible to install any specified version from conda-forge
6:  493f366c3 ! 5:  e3b69a0b8 BF: use old 7.20190819 git-annex from conda-forge
    @@ .travis.yml: env:
          # get ahead in PATH to not shadow travis' python
     -    - _DL_ANNEX_INSTALL_SCENARIO=conda-forge-last
     +    - _DL_ANNEX_INSTALL_SCENARIO="conda-forge-last 7.20190819"
    - 
    - 
    - before_install:
    +   matrix:
    + ##1    - DATALAD_REPO_VERSION=6
    +     - DATALAD_REPO_VERSION=5
7:  00ee6a82e = 6:  7bd1e6760 BF: travis - store adjusted PATH export within ~/.bashrc
-:  --------- > 7:  593094ffa TST: travis: Reposition PATH adjustment for git-annex
```

</details>

**Comment by kyleam** [comment]
There are three jobs with tools/ci/install-annex.sh errors, and one job that has a single failing test (perhaps due to brittle commit counting, but I haven't taken a closer look).

https://travis-ci.org/github/datalad/datalad/builds/703721783

**Comment by yarikoptic** [comment]
> There are three jobs with tools/ci/install-annex.sh errors

Chicken-egg problem somewhat around TMPDIR. I have pushed 2 commits which reorder things a bit.  Not yet sure if they would work.  Comments provide some additional information.  Feel free to drop if you see them not needed etc.

File: .travis.yml [hunk 1]
```diff
@@ -220,34 +217,40 @@ before_install:
...
   - travis_retry sudo eatmydata apt-get install xz-utils
   - travis_retry sudo eatmydata apt-get install shunit2
   # for metadata support
   - travis_retry sudo eatmydata apt-get install exempi
+  # Configure TMPDIR before trying install git-annex -- script might use it
+  - if [[ "${TMPDIR:-}" =~ .*/sym\ link ]]; then echo "Symlinking $TMPDIR"; ln -s /tmp "$TMPDIR"; fi
+  - if [[ "${TMPDIR:-}" =~ .*/d\ i\ r ]]; then echo "mkdir $TMPDIR"; mkdir -p "$TMPDIR"; fi
+  - if [[ "${TMPDIR:-}" =~ .*/nfsmount ]]; then echo "mkdir $TMPDIR"; mkdir -p "$TMPDIR" "${TMPDIR}_"; echo "/tmp/nfsmount_ localhost(rw)" | sudo bash -c 'cat - > /etc/exports'; sudo apt-get install -y nfs-kernel-server; sudo exportfs -a; sudo mount -t nfs localhost:/tmp/nfsmount_ /tmp/nfsmount; fi
+  # Install git-annex
+  - OLD_PATH="$PATH"
```
Comment (yarikoptic) [inline .travis.yml:232]: Now I am thinking that we might just want to add options (e.g. `--symlink-under=PATH`) to the install script to install `git-annex` and `git-annex-shell` into `/usr/local/bin` which we would use in such cases and avoid the need for `source` (which should still be "an option" for other testing deployments)

**Comment by yarikoptic** [comment]
Those two resolved some problems. We have only two builds errorring out in miniconda installation script not tollerating spaces in the path well:
```
I: top directory /var/tmp/sym link/ga-c9VY12C
...
2020-07-01 03:39:31 (193 MB/s) - ‘/var/tmp/sym link/ga-c9VY12C/Miniconda3-latest-Linux-x86_64.sh’ saved [88867207/88867207]
ERROR: did not recognize option 'link/ga-c9VY12C/miniconda', please try -h
tools/ci/install-annex.sh: line 88: /var/tmp/sym link/ga-c9VY12C/miniconda/bin/conda: No such file or directory
The command "eval source tools/ci/install-annex.sh ${_DL_ANNEX_INSTALL_SCENARIO}" failed and exited with 127 during .
```
and conda installer would have failed anyways: filed https://github.com/conda/constructor/issues/347 .  So, great that we found this issue, but I guess to resolve it ATM we would need to proceed with what I commented in commit 83d51abfcf3854ceda4a3b496b33bc9cdcc4a388 :
 
   Which gets me thinking that may be we should not just export TMPDIR straight
    in the environment but rather e.g. as TMPDIR_ and then assign to TMPDIR after
    they are setup, so git annex installation does not end up on some odd TMPDIR
    to be tested.

**Comment by kyleam** [comment]
[ see you just posted, but since I just typed this out and it seems like you want to go a different route ]

The `TMPDIR="/var/tmp/sym link"` jobs still fail in the setup stage.  The error happens in the argument parsing miniconda install script.  We could get by that particular failure by switching to the temporary directory in a subshell, but the mininconda script will just fail later on.

I'm not sure what you have in mind for this script, but I don't think see much value in worrying about $TMPDIR.  I think we should just install in either the current directory or /tmp (which should work for the context we need it).

<details>
<summary>later failure</summary>

```
_miniconda_script=Miniconda3-latest-Linux-x86_64.sh
TMPDIR="/var/tmp/sym link"
_TMPDIR=$(mktemp -d "${TMPDIR:-/tmp}/ga-XXXXXXX")
wget -O "$_TMPDIR/${_miniconda_script}" "${ANACONDA_URL:-https://repo.anaconda.com/miniconda/}${_miniconda_script}"
(
    cd "$_TMPDIR"
    HOME="$_TMPDIR"
    bash "${_miniconda_script}" -b -p miniconda
)

```

```
[... 97 lines ...]
Preparing transaction: done
Executing transaction: done
rm: cannot remove '/var/tmp/sym': Is a directory
rm: cannot remove '/var/tmp/sym': Is a directory
Miniconda3-latest-Linux-x86_64.sh: line 449: /var/tmp/sym: Is a directory
```
</details>


https://travis-ci.org/github/datalad/datalad/jobs/703768393#L779

File: .travis.yml [hunk 2]
```diff
@@ -220,34 +217,40 @@ before_install:
...
+  - eval source tools/ci/install-annex.sh ${_DL_ANNEX_INSTALL_SCENARIO}
```
Comment (kyleam) [inline .travis.yml:233]: Why is `eval` needed?

Comment (yarikoptic) [reply .travis.yml:233]: may be it is not? I think I added it assuming that if I want multiple arguments in `_DL_ANNEX_INSTALL_SCENARIO`...  may be I overdone it

**Comment by yarikoptic** [comment]
wooohoo -- it is green now!  Added some TODOs to the top description

**Comment by kyleam** [comment]
> verify that actually setting TMPDIR was in effect (e.g. pidlock was engaged in nfs based test case)

Comparing this PR's `/tmp/nfsmount` [job][0] with a recent [job from maint][0] suggests to me that TMPDIR is being adjusted as intended: same number of skipped tests (perhaps not relevant, since I don't spot any NFS-specific SkipTest calls in the code base) and the same traceback during cleanup about "OSError: [Errno 16] Device or resource busy: '.nfs*".

Also, in this `_DL_TMPDIR="/var/tmp/sym link"` [job][2], that location can be spotted in the logs.

[0]: https://travis-ci.org/github/datalad/datalad/jobs/707662863#L1561
[1]: https://travis-ci.org/github/datalad/datalad/jobs/707619311#L1376
[2]: https://travis-ci.org/github/datalad/datalad/jobs/707662861#L5219

**Comment by kyleam** [comment]
> move tune up of ~/.bashrc inside that helper (as cmdline option).  I think we should avoid such actions to be done in .travis.yml, so we could reuse script across CIs etc.

I'll adjust the script.

**Comment by kyleam** [comment]
Moved .bashrc path adjustment to install-annex.sh.  Travis still passed. 

@yarikoptic I'll leave the merge to you in case you'd like to make tweaks to the additions or if the [reasoning above][0] doesn't convince you that this script didn't break the handling of TMPDIR.

[0]: https://github.com/datalad/datalad/pull/4640#issuecomment-658245569

File: tools/ci/install-annex.sh [hunk 3]
```diff
@@ -33,12 +33,40 @@ function setup_neurodebian_devel() {
...
+                neurodebian|neurodebian-devel|snapshot|datalad-extensions-build)
+                    ;;
+                conda-forge|conda-forge-last)
+                    if [ -n "$1" ]; then
+                        _conda_annex_version="=$1"
+                        shift
+                    fi
+                    ;;
+                deb-url)
+                    url="${1?deb-url scheme requires URL}"
```
Comment (yarikoptic) [inline tools/ci/install-annex.sh:78]: oh didn't know about this shell trick.

File: tools/ci/install-annex.sh [hunk 4]
```diff
@@ -49,9 +49,9 @@ scenario="conda-forge"
 url=
 while [ $# != 0 ]; do
     case "$1" in
-        --help)
+        -*)
             _usage
-            exit 0
+            exit 1
```
Comment (yarikoptic) [inline tools/ci/install-annex.sh:63]: hm, so `--help` would exit with 1? IMHO suboptimal

**Comment by kyleam** [comment]
> I will add "special" handling back for --help to exit with 0 and merge/push

As you wish.  Not my preference, but also doesn't matter.

**Comment by yarikoptic** [comment]
Why should it be an error exit upon --help? All other tools (git, ls) exit normally, so it is IMHO not a matter of preference even

**Pull request merged**